Fix Pathfollow direction and add Z forward option

TokageItLab opened the pull request:
Fixes #53837.

The direction is changed, but some countermeasures are planned in the near future. See #72843, #72795 and #72753.

~~As long as #72843 will be merged, the change about the direction is not breaks compat. However, the following method changes are breaks compat.~~ merged into this PR.

---
PathFollow2D:
- ~~Rename: `rotates` `set_rotates` `is_rotating` -> `rotation_enabled` `set_rotation_enabled` `is_rotation_enabled`~~
- Remove: `lookahead` the DEAD PROPERTY💀

PathFollow2D/3D:
- ~~Rename: `cubic_interp` `set_cubic_interpolation` `get_cubic_interpolation` -> `cubic_interpolation_enabled` `set_cubic_interpolation_enabled` `is_cubic_interpolation_enabled`~~

~~But, to avoid breaking the scene, old methods is made migratable as deprecated in `_set()` `_get()`.~~

fire commented:
I am supporting this, and this is a local change.

TokageItLab committed:
Fix Pathfollow direction

Co-authored-by: aaronfranke <arnfranke@yahoo.com>

TokageItLab commented:
Rebased, and integrated https://github.com/godotengine/godot/pull/72843.

aaronfranke commented on scene/3d/path_3d.cpp:
```diff
@@ -191,6 +191,9 @@ void PathFollow3D::_update_transform(bool p_update_xyz_rot) {
 	} else {
 		t = c->sample_baked_with_rotation(progress, cubic, false);
+		if (z_forward) {
+			t.basis = Basis(t.basis.get_column(1).normalized(), Math_PI) * t.basis;
```
Performing this conversion by using the axis-angle Basis constructor is not the ideal solution. I think this code will perform an equivalent operation:

```suggestion
			t.basis *= Basis::from_scale(-1.0, 1.0, -1.0);
```

(Linear algebra trick: A negative scale on two axes performs a 180 degree rotation on the plane formed by those axes. And we are doing this from the right so it will be in local space.)

TokageItLab commented:
Rebased to follow #76082. The prerequisite is to merge #76082 first, since it contains links to class references added in #76082.

Also, for the same reasons as https://github.com/godotengine/godot/pull/74190#issuecomment-1546352973, I have changed to keep the bound method names for compatibility although the function names have been changed for consistency in cpp.

aaronfranke commented on doc/classes/PathFollow2D.xml:
```diff
@@ -18,9 +18,6 @@
 		<member name="h_offset" type="float" setter="set_h_offset" getter="get_h_offset" default="0.0">
 			The node's offset along the curve.
 		</member>
-		<member name="lookahead" type="float" setter="set_lookahead" getter="get_lookahead" default="4.0">
```
Isn't this compat-breaking?

akien-mga replied:
Technically it does break compatibility: GDScript/C# code that refers to it will break, and I think this might impact the GDExtension ABI even for code that doesn't use it (but not sure about that, CC @godotengine/gdextension).

Since the property really did nothing, for a minor release it's acceptable I think, as long as we properly document it as a removal in the CHANGELOG.md (so I'll add the `breaks compat` label so we can keep track of it).

TokageItLab replied:
Okay thanks, I marked breaks compat.

aaronfranke commented on doc/classes/Vector3.xml:
```diff
@@ -404,11 +404,29 @@
 		</constant>
 		<constant name="FORWARD" value="Vector3(0, 0, -1)">
-			Forward unit vector. Represents the local direction of forward, and the global direction of north.
+			Forward unit vector. Represents the local direction of forward, and the global direction of north. Keep in mind that, in Godot, the forward vector is generally opposed to where imported 3D assets are facing. Use [constant Vector3.MODEL_FRONT] and similar constants to manipulate transforms in imported 3D asset space.
```
```suggestion
			Forward unit vector. Represents the local direction of forward, and the global direction of north. Keep in mind that the forward direction for lights, cameras, etc is different from 3D assets like characters, which face towards the camera by convention. Use [constant Vector3.MODEL_FRONT] and similar constants when working in 3D asset space.
```

TokageItLab replied:
I integrated this change in https://github.com/godotengine/godot/pull/76082.

aaronfranke commented on scene/3d/path_3d.cpp:
```diff
@@ -230,11 +233,11 @@ void PathFollow3D::_notification(int p_what) {
 }
 
-bool PathFollow3D::get_cubic_interpolation() const {
+bool PathFollow3D::is_cubic_interpolation_enabled() const {
```
These changes seem unrelated to adding a new model front mode to `look_at`. I'm not sure why not split these changes into a separate PR.

TokageItLab replied:
I talked to @akien-mga earlier and got his agreement that these method names are bad, but he seems to have a terrible aversion to changing method names in 4.x, so I decided to change only the cpp function names. same as https://github.com/godotengine/godot/pull/74190#issuecomment-1546352973. Binding method names have not been changed. If he does not like it, I will split it.

aaronfranke reviewed (commented):
In general I support this PR. Adding this option is a good idea, and naming it model front is a very good sensible choice.

I was initially opposed to adding new constants when they had the `GLTF_`* names, but I now after discussion and with the new names I think they are very good and sensible. The constants in this PR are very descriptive and are self-consistent with each other.

aaronfranke commented on core/math/transform_3d.cpp:
```diff
@@ -77,7 +77,7 @@ void Transform3D::rotate_basis(const Vector3 &p_axis, real_t p_angle) {
 }
 
-Transform3D Transform3D::looking_at(const Vector3 &p_target, const Vector3 &p_up) const {
+Transform3D Transform3D::looking_at(const Vector3 &p_target, const Vector3 &p_up, bool p_use_model_front) const {
```
I compared this PR with #75875 and I found a bug, you forgot to use this parameter.

aaronfranke commented on scene/3d/node_3d.cpp:
```diff
@@ -908,20 +908,24 @@ void Node3D::set_identity() {
 	ERR_FAIL_COND_MSG(p_up.cross(p_target - p_pos).is_zero_approx(), "Up vector and direction between node origin and target are aligned, look_at() failed.");
 
-	Transform3D lookat = Transform3D(Basis::looking_at(p_target - p_pos, p_up), p_pos);
+	Vector3 forward = p_target - p_pos;
+	if (p_use_model_front) {
+		forward = -forward;
+	}
+	Transform3D lookat = Transform3D(Basis::looking_at(forward, p_up), p_pos);
 	Vector3 original_scale = get_scale();
 	set_global_transform(lookat);
```
You can just pass `p_use_model_front` into `Basis::looking_at`. I think the code would be more readable like I suggest here, and it avoids a double-negation in some situations:

```suggestion
	Vector3 forward = p_target - p_pos;
	Basis lookat_basis = Basis::looking_at(forward, p_up, p_use_model_front);
	Vector3 original_scale = get_scale();
	set_global_transform(Transform3D(lookat_basis, p_pos));
```

aaronfranke reviewed (commented):
I updated my PR #75875 to be identical to the Basis/Transform3D/Node3D changes in this PR, excluding the path/curve changes, with the below review comments applied, and with this line added to each method's documentation:

`If [param use_model_front] is [code]true[/code], the +Z axis (asset front) is treated as forward (implies +X is left) and points toward the [param target] position. By default, the -Z axis (camera forward) is treated as forward (implies +X is right).`

(I also marked you as author since I started by copying the contents of this PR exactly)

aaronfranke reviewed (approved):
I approve of these changes. In the event that the path/curve stuff is desired to be split off from the look_at stuff (pending decision from @akien-mga), I have this PR available that is mostly a subset of this one: #75875

akien-mga commented:
> For the record, renaming method arguments is probably breaking compat too (for C# at least?)

That's true, but I think we've already renamed a lot of method arguments in 4.1 already. If we want to preserve this level of compatibility for C# (I personally think it's too demanding), we'd need another dedicated compatibility system for it.

akien-mga commented:
Thanks!

YuriSizov commented:
Yeah, I agree. Just mentioning it so it's clear to anyone looking through the changes.

raulsntos commented:
Renaming arguments in C# is a source breaking change, not a binary breaking change. This means, if you are using a library that was built against 4.0 that library will still continue to work in 4.1 without having to recompile it. A source breaking change is only an issue for the application/game code but it's usually easy to fix.

In this case, the code will break only if you are explicitly using the argument names:

```csharp
void MyMethod(int one, int two, int three) { ... }

// This method call will not break if the arguments of MyMethod are renamed.
MyMethod(1, 2, 3);

// This method call will break if one of the arguments of MyMethod is renamed.
MyMethod(one: 1, two: 2, three: 3);
```

I'm more worried about binary breaking changes because it's a lot more disruptive.
